src/algebra/gcd_domain.lean

```diff
@@ -27,41 +27,108 @@ attribute [simp] norm_unit_coe_units norm_unit_zero norm_unit_mul
+variables {α}
+
+lemma normalize_coe_units (u : units α) : normalize (u : α) = 1 :=
+by rw [normalize, norm_unit_coe_units, ← units.coe_mul, mul_inv_self, units.coe_one]
+
+theorem normalize_mul (x y : α) : normalize (x * y) = normalize x * normalize y :=
+classical.by_cases (λ hx : x = 0, by rw [hx, zero_mul, normalize_zero, zero_mul]) $ λ hx,
+classical.by_cases (λ hy : y = 0, by rw [hy, mul_zero, normalize_zero, mul_zero]) $ λ hy,
+by simp only [normalize, norm_unit_mul hx hy, units.coe_mul]; simp only [mul_assoc, mul_left_comm y]
+
+@[simp] lemma normalize_eq_zero {x : α} : normalize x = 0 ↔ x = 0 :=
+⟨λ hx, (associated_zero_iff_eq_zero x).1 $ hx ▸ associated_normalize, by rintro rfl; exact normalize_zero α⟩
+
+@[simp] lemma normalize_eq_one {x : α} : normalize x = 1 ↔ is_unit x :=
+⟨λ hx, is_unit_iff_exists_inv.2 ⟨_, hx⟩, λ ⟨u, hu⟩, hu.symm ▸ normalize_coe_units u⟩
+
 theorem norm_unit_mul_norm_unit (a : α) : norm_unit (a * norm_unit a) = 1 :=
 classical.by_cases (assume : a = 0, by simp only [this, norm_unit_zero, zero_mul]) $
-  assume h, by rw [norm_unit_mul _ _ h (units.ne_zero _), norm_unit_coe_units, mul_inv_eq_one]
-
-theorem associated_of_dvd_of_dvd : ∀{a b : α}, a ∣ b → b ∣ a → ∃u:units α, a * u = b
-| a b ⟨c, rfl⟩ ⟨d, hd⟩ :=
-  classical.by_cases (assume : c = 0, ⟨1, by simp only [zero_mul, mul_zero, *] at *⟩) $ assume hc : c ≠ 0,
-  classical.by_cases (assume : a = 0, ⟨1, by simp only [zero_mul, *] at *⟩) $ assume ha : a ≠ 0,
-  have a * 1 = a * (c * d), by rwa [mul_one, ← mul_assoc],
-  have 1 = c * d, from eq_of_mul_eq_mul_left ha this,
-  ⟨units.mk_of_mul_eq_one c d this.symm, rfl⟩
-
-theorem mul_norm_unit_eq_mul_norm_unit {a b : α}
-  (hab : a ∣ b) (hba : b ∣ a) : a * norm_unit a = b * norm_unit b :=
+  assume h, by rw [norm_unit_mul h (units.ne_zero _), norm_unit_coe_units, mul_inv_eq_one]
+
+theorem normalize_idem (x : α) : normalize (normalize x) = normalize x :=
+by rw [normalize, normalize, norm_unit_mul_norm_unit, units.coe_one, mul_one]
+
+theorem normalize_eq_normalize {a b : α}
+  (hab : a ∣ b) (hba : b ∣ a) : normalize a = normalize b :=
 begin
-  rcases associated_of_dvd_of_dvd hab hba with ⟨u, rfl⟩,
+  rcases associated_of_dvd_dvd hab hba with ⟨u, rfl⟩,
   refine classical.by_cases (by rintro rfl; simp only [zero_mul]) (assume ha : a ≠ 0, _),
   suffices : a * ↑(norm_unit a) = a * ↑u * ↑(norm_unit a) * ↑u⁻¹,
-    by simpa only [mul_assoc, norm_unit_mul a u ha (units.ne_zero _), norm_unit_coe_units],
+    by simpa only [normalize, mul_assoc, norm_unit_mul ha u.ne_zero, norm_unit_coe_units],
   calc a * ↑(norm_unit a) = a * ↑(norm_unit a) * ↑u * ↑u⁻¹:
       (units.mul_inv_cancel_right _ _).symm
     ... = a * ↑u * ↑(norm_unit a) * ↑u⁻¹ : by rw mul_right_comm a
 end
 
-theorem dvd_antisymm_of_norm {a b : α}
-  (ha : norm_unit a = 1) (hb : norm_unit b = 1) (hab : a ∣ b) (hba : b ∣ a) :
+@[simp] lemma normalize_eq_normalize_iff {x y : α} : normalize x = normalize y ↔ x ∣ y ∧ y ∣ x :=
```

Comment:
this doesn't look like it should be a simp lemma either

Reply:
done